src/content/docs/dev/general-message-passing/sui/gmp-tutorial.mdx

```diff
@@ -0,0 +1,918 @@
+gateway.send_message(message_ticket);
+```
+
+The complete `send_message()` function should look like this:
+
+```rust
+public fun send_message(
+    singleton: &Singleton,
+    gateway: &Gateway,
+    gas_service: &mut GasService,
+    destination_chain: String,
+    destination_address: String,
+    payload: vector<u8>,
+    refund_address: address,
+    coin: Coin<SUI>,
+    params: vector<u8>,
+) {
+    // Create a new message ticket
+    let message_ticket = gateway::prepare_message(
+        &singleton.channel,
+        destination_chain,
+        destination_address,
+        payload,
+    );
+
+    // Pay for the gas
+    gas_service.pay_gas(
+        &message_ticket,
+        coin,
+        refund_address,
+        params,
+    );
+
+    // Dispatch the cross-chain message
+    gateway.send_message(message_ticket);
+}
+```
+
+### Receive Message
+
+With the package now able to send cross-chain messages, it would be great to also receive messages from other blockchains. This can be done by creating an `execute()` function that will be triggered by our Axelar's relaying infrastructure for any incoming message. The full function signature for the `execute()` function will require two parameters:
+
+1. `call: ApprovedMessage`: The [approved message](/dev/general-message-passing/sui/gmp-contracts/#approvedmessage) that was sent from the source chain
+1. `singleton: &mut Singleton`: A mutable reference to the `singleton` resource so the package can consume messages from its `channel` 
+
+```rust
+public fun execute(
+    call: ApprovedMessage, 
+    singleton: &mut Singleton
+) {}
+```
+
+Within the `execute()` function you can then consume the message by triggering the [consume_approved_message()](/dev/general-message-passing/sui/gmp-contracts/#consume-approved-message) function that is [defined](https://github.com/axelarnetwork/axelar-cgp-sui/blob/main/move/axelar_gateway/sources/channel.move#L85) in the `Channel` module. Once the message is approved you can `emit` a custom event indicating that the execution has gone through successfully. 
+
+Your complete `execute()` function should look like this:
+
+```rust
+public fun execute(
+    call: ApprovedMessage, 
+    singleton: &mut Singleton
+) {
+    // Consume the message
+    let (_, _, _, payload) = singleton.channel.consume_approved_message(call);
+
+    // Emit Executed event
+    event::emit(Executed { data: payload });
+}
+```
+
+This will lead to an error however as you have not yet defined your event, you can do so up top as follows. Note unlike the `Singleton` struct that had a `key` allowing it to be its own on-chain object, the `Executed` struct has a `copy` + `drop` [abilities](https://docs.sui.io/references/sui-api/sui-graphql/reference/types/enums/move-ability) so the event data can be duplicated into the transaction’s event log and then discarded by off-chain indexers. The event itself simply takes a raw payload of bytes that it will be receiving in the cross-chain call.
```

Comment:
```suggestion
This will lead to an error, however, as you have not yet defined your event, you can do so up top as follows. Note unlike the `Singleton` struct that had a `key` allowing it to be its own on-chain object, the `Executed` struct has a `copy` + `drop` [abilities](https://docs.sui.io/references/sui-api/sui-graphql/reference/types/enums/move-ability) so the event data can be duplicated into the transaction’s event log and then discarded by off-chain indexers. The event takes a raw payload of bytes that it will receive in the cross-chain call.
```